src/org/mirah/jvm/mirrors/mirror_type_system.mirah

```diff
@@ -744,6 +751,46 @@ class MirrorTypeSystem implements TypeSystem
     c = a.widen(b)
     puts c
   end
+
+  def register_extensions:void
+    boot_class_loader = MirrorTypeSystem.class.getClassLoader
+    register_extensions boot_class_loader
+    
+    compile_class_loader = ClassLoader(@context[ClassLoader])
+    
+    if compile_class_loader != boot_class_loader
+      register_extensions compile_class_loader
+    end
+  end
+  # use java service SPI to load all extensions registrations from context classloader
+  def register_extensions(class_loader:ClassLoader):void    
+    return unless class_loader
+    services = ServiceLoader.load(ExtensionsProvider.class, class_loader)
+    type_system = self
+    iter = Iterable(services).iterator
+    while iter.hasNext
+      provider = ExtensionsProvider(iter.next)
+      provider.register(type_system)
+    end
+  rescue Error => ex
+    puts "Unable to load regtistration service: #{ex}"
+  end
+
+  # ExtensionsService implementation
+  def macro_registration(clazz:Class):void
+    anno = clazz.getAnnotation(ExtensionsRegistration.class)
+    macro_clazz = @context[ClassLoader].loadClass("#{clazz.getName}$Extensions") rescue nil
+    # different ways for extensions annotations
+    macro_clazz = clazz unless macro_clazz
+    type_system = self
+    unless anno.nil?
+      anno.value.each do |class_name|
+        type_system.extendClass(class_name, macro_clazz)
+      end
+    end
+  rescue Error => ex
```

Comment:
I feel like this should rescue Exception, unless there's a particular Error subclass this might raise.